exp/runtime/api/v1beta1/extensionconfig_types.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+// ANCHOR: ExtensionConfigSpec
+
+// ExtensionConfigSpec defines the desired state of ExtensionConfig.
+type ExtensionConfigSpec struct {
+	// ClientConfig defines how to communicate with ExtensionHandlers.
+	ClientConfig ClientConfig `json:"clientConfig"`
+
+	// NamespaceSelector decides whether to run the webhook on an object based
+	// on whether the namespace for that object matches the selector.
+	// Default to the empty LabelSelector, which matches everything.
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+}
+
+// ClientConfig contains the information to make a client
+// connection with an ExtensionHandler.
+type ClientConfig struct {
+	// URL gives the location of the ExtensionHandler, in standard URL form
+	// (`scheme://host:port/path`). Exactly one of `url` or `service`
+	// must be specified.
+	//
+	// The `host` should not refer to a service running in the cluster; use
+	// the `service` field instead.
+	//
+	// The scheme should be "https"; the URL should begin with "https://".
+	// "http" is supported for insecure development purposes only.
+	//
+	// A path is optional, and if present may be any string permissible in
+	// a URL. If a path is set it will be used as prefix and the hook-specific
+	// path will be appended.
+	//
+	// Attempting to use a user or basic auth e.g. "user:password@" is not
+	// allowed. Fragments ("#...") and query parameters ("?...") are not
+	// allowed either.
+	//
+	// +optional
+	URL *string `json:"url,omitempty"`
+
+	// Service is a reference to the Kubernetes service for the ExtensionHandler.
+	// Either `service` or `url` must be specified.
+	//
+	// If the ExtensionHandler is running within a cluster, then you should use `service`.
+	//
+	// +optional
+	Service *ServiceReference `json:"service,omitempty"`
+
+	// CABundle is a PEM encoded CA bundle which will be used to validate the ExtensionHandler's server certificate.
+	// +optional
+	CABundle []byte `json:"caBundle,omitempty"`
```

Comment:
If this is bounded by any minimum or maximum value, it can be helpful to add a kubebuilder validation for this

Reply:
I added an issue for kubebuilder validation here #6398 I'm not sure if it's a good  to have a separate space to discuss it. For this PR I think it would be reasonably easy to add it, but I think we need to decide whether it's a good idea to use this validation at all in our APIs and whether to bring APIs that don't use it into alignment going forward.

Reply:
I have [commented on 6398](https://github.com/kubernetes-sigs/cluster-api/issues/6398#issuecomment-1094898385), but I do not think we should put this decision in the critical path for this PR, because it seems there is a foundational discussion about how we test defaulting/validation and conversion to be addressed before changing our current approach.